### PR TITLE
core: Change string list representation so we can distinguish empty, single element lists

### PR DESCRIPTION
**Pull request by phinze**
This encapsulates and enhances the internal representation of lists within terraform's core.

To work around limitations of the current architecture, lists need to be "flattened" into a string as they're passed around the config interpolation.

Given a list `["foo", "bar", "baz"]`, the previous format for these strings was: `foo(special delimiter)bar(special delimiter)baz`.

This format made it impossible to distinguish between:

 * `["foo"]` and `"foo"` 
 * `[]` and `""`
 * `[""]` and `""`

Here's the new way we use the "string list delimiter" (SLD) as described in a comment:

```
// It plays two semantic roles:
//   * It introduces a list
//   * It terminates each element
//
// Example representations:
// []             => SLD
// [""]           => SLDSLD
// [" "]          => SLD SLD
// ["foo"]        => SLDfooSLD
// ["foo", "bar"] => SLDfooSLDbarSLD
// ["", ""]       => SLDSLDSLD
```

### CODE REVIEW

File: config/interpolate_funcs.go [hunk 1]
```diff
@@ -54,12 +54,13 @@ func interpolationFuncConcat() ast.Function {
...
 
-				if strings.Contains(argument, InterpSplitDelim) {
+				if IsStringList(argument) {
 					isDeprecated = false
+					finalList = append(finalList, StringList(argument).Slice()...)
+				} else {
+					finalList = append(finalList, argument)
 				}
 
-				finalList = append(finalList, argument)
```
Comment (phinze) [inline config/interpolate_funcs.go]: This was implicitly relying on the fact that the old representation of StringLists allowed them to be arbitrarily combined with bare elements.

For example, previously: `["foo", "barSLDbaz", "qux"]` would make it through the `strings.Join` as a valid list. Now we need to respect the encapsulation of the type and handle it explicitly.

File: config/interpolate_funcs.go [hunk 2]
```diff
@@ -131,11 +132,16 @@ func interpolationFuncFormatList() ast.Function {
 				if !ok {
 					continue
 				}
-				parts := strings.Split(s, InterpSplitDelim)
-				if len(parts) == 1 {
+				if !IsStringList(s) {
 					continue
 				}
+
+				parts := StringList(s).Slice()
```
Comment (phinze) [inline config/interpolate_funcs.go]: This was previously relying on the fact that a bare string would have no problem being split. Need to handle it explicitly now.

File: config/interpolate_funcs.go [hunk 3]
```diff
@@ -246,7 +249,9 @@ func interpolationFuncSplit() ast.Function {
 		ArgTypes:   []ast.Type{ast.TypeString, ast.TypeString},
 		ReturnType: ast.TypeString,
 		Callback: func(args []interface{}) (interface{}, error) {
-			return strings.Replace(args[1].(string), args[0].(string), InterpSplitDelim, -1), nil
+			sep := args[0].(string)
+			s := args[1].(string)
+			return NewStringList(strings.Split(s, sep)).String(), nil
```
Comment (phinze) [inline config/interpolate_funcs.go]: Can't cheat with a `Replace` here anymore. :grinning:

File: config/interpolate_funcs.go [hunk 4]
```diff
@@ -292,7 +297,7 @@ func interpolationFuncElement() ast.Function {
 					"invalid number for index, got %s", args[1])
 			}
 
-			v := list[index%len(list)]
+			v := list.Element(index)
```
Comment (phinze) [inline config/interpolate_funcs.go]: Easiest to just push this down to the StringList type - with the mod behavior living down there.

File: config/interpolate_funcs_test.go [hunk 5]
```diff
@@ -219,23 +190,18 @@ func TestInterpolateFuncFormatList(t *testing.T) {
...
 				`${formatlist("%s=%2s", split(",", "A,B,C,D"), split(",", "1,2,3"))}`,
 				nil,
 				true,
 			},
+			// Works with lists of length 1 [GH-2240]
+			{
+				`${formatlist("%s.id", split(",", "demo-rest-elb"))}`,
+				NewStringList([]string{"demo-rest-elb.id"}).String(),
+				false,
+			},
```
Comment (phinze) [inline config/interpolate_funcs_test.go]: This tiny detail was the instigator of this whole rabbit hole, heh.

File: config/interpolate_funcs_test.go [hunk 6]
```diff
@@ -42,74 +41,46 @@ func TestInterpolateFuncConcat(t *testing.T) {
...
 			{
 				`${concat("a", split(",", "b,c"))}`,
-				fmt.Sprintf(
-					"%s%s%s%s%s",
-					"a",
-					InterpSplitDelim,
-					"b",
-					InterpSplitDelim,
-					"c"),
+				NewStringList([]string{"a", "b", "c"}).String(),
```
Comment (phinze) [inline config/interpolate_funcs_test.go]: feelsgoodman.gif :+1:

File: config/interpolate_funcs_test.go [hunk 7]
```diff
@@ -387,46 +351,38 @@ func TestInterpolateFuncSplit(t *testing.T) {
...
 			{
+				`${split(",", "")}`,
+				NewStringList([]string{""}).String(),
+				false,
+			},
+
+			{
 				`${split(",", "foo")}`,
-				"foo",
+				NewStringList([]string{"foo"}).String(),
```
Comment (phinze) [inline config/interpolate_funcs_test.go]: Important detail that this is definitely output as a list now.

File: config/interpolate_walk.go [hunk 8]
```diff
@@ -10,10 +10,6 @@ import (
 	"github.com/mitchellh/reflectwalk"
 )
 
-// InterpSplitDelim is the delimeter that is looked for to split when
-// it is returned.
-const InterpSplitDelim = `B780FFEC-B661-4EB8-9236-A01737AD98B6`
```
Comment (phinze) [inline config/interpolate_walk.go]: :wave:

File: config/interpolate_walk.go [hunk 9]
```diff
@@ -148,8 +144,8 @@ func (w *interpolationWalker) Primitive(v reflect.Value) error {
 		// set if it is computed. This behavior is different if we're
 		// splitting (in a SliceElem) or not.
 		remove := false
-		if w.loc == reflectwalk.SliceElem {
-			parts := strings.Split(replaceVal, InterpSplitDelim)
+		if w.loc == reflectwalk.SliceElem && IsStringList(replaceVal) {
+			parts := StringList(replaceVal).Slice()
```
Comment (phinze) [inline config/interpolate_walk.go]: Another example of a place where we were leaning on the fact that a bare string could survive a `Split`. Need to fall through to the next case for plain strings.

File: config/string_list.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+//   * It terminates each element
+//
+// Example representations:
+// []             => SLD
+// [""]           => SLDSLD
+// [" "]          => SLD SLD
+// ["foo"]        => SLDfooSLD
+// ["foo", "bar"] => SLDfooSLDbarSLD
+// ["", ""]       => SLDSLDSLD
+const stringListDelim = `B780FFEC-B661-4EB8-9236-A01737AD98B6`
```
Comment (phinze) [inline config/string_list.go]: This can be private now. :information_desk_person:

File: config/string_list.go [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+		return []string{""}
+	}
+
+	// strip empty elements generated by leading and trailing delimiters
+	return parts[1 : len(parts)-1]
+}
+
+func (sl StringList) String() string {
+	return string(sl)
+}
```
Comment (phinze) [inline config/string_list.go]: Decided to go with an explicit function to "convert" to a string. Felt right but would appreciate commentary on the most go-like thing to do here.

File: terraform/context_test.go [hunk 12]
```diff
@@ -1672,29 +1672,6 @@ func TestContext2Plan_provider(t *testing.T) {
...
-	if err != nil {
-		t.Fatalf("err: %s", err)
-	}
-
-	actual := strings.TrimSpace(plan.String())
-	expected := strings.TrimSpace(testTerraformPlanVarMultiCountOneStr)
-	if actual != expected {
-		t.Fatalf("bad:\n%s", actual)
-	}
-}
```
Comment (phinze) [inline terraform/context_test.go]: I talked about this in the commit message, but this old test was actually invalid config and its idea is covered in `TestContext2Apply_multiVar`.

File: config/interpolate_funcs_test.go (hunk 6 above)
Comment (mitchellh) [inline config/interpolate_funcs_test.go]: :D

**Comment by mitchellh** [comment]
Surprisingly simple. LGTM!

**Pull request merged**